graphs/possible_bipartition.py

```diff
@@ -1,5 +1,22 @@
 # Can be used for BFS
-from collections import deque 
+from collections import deque
+# This is DFS
+def has_no_bad_cycle(cycle_counter_list, key, dislikes):
+    # exits if the key is present in list => there is a cycle
+    # AND
+    # if the len of list(path) is odd => the cycle is NOT ok
+    if len(cycle_counter_list) % 2 != 0 and cycle_counter_list[0] == key:
+        return False
+    
+    # exits if the key is present in list => there is a cycle
+    if key in cycle_counter_list:
+        return True
+
+    for value in dislikes[key]:
+        if has_no_bad_cycle(cycle_counter_list + [key], value, dislikes) == False:
```

Comment:
Add the current dog to `cycle_counter_list` and call the helper function on the enemy dog to see if the current dog and enemy dogs can be partitioned or not.

This is what makes your implementation DFS! It's using a stack (the call stack via recursion) instead of a queue.